File: jsds/栈/Stack.js
import SinglyLinkedList from '../链表/SinglyLinkedList.js'
/**
 push(element)：添加一个（或几个）新元素到栈顶
 pop()：移除栈顶的元素，同时返回被移除的元素
 peek()：返回栈顶的元素，不对栈做任何修改
 isEmpty()：如果栈里没有任何元素就返回true，否则返回false
 clear()：移除栈里的所有元素
 size()：返回栈里的元素个数

 A,B,C,D 进栈
 先进后出   D C B A
 */
export default class Stack {
  constructor(){
    this.linkedList = new SinglyLinkedList()
  }

  push (value) {
    this.linkedList.prepend(value)
    return this
  }

  pop () {
    const removeHead = this.linkedList.deleteHead()
    return removeHead ? removeHead : null
  }

  peek () {
    if (this.isEmpty()) {
      return null
    }
    return this.linkedList.head
  }

  isEmpty () {
    return !this.linkedList.head
  }

  size () {
    return this.linkedList.size()
  }

  clear () {
    this.linkedList = null
  }

  toArray () {
    return this.linkedList.toArray()
  }
}

File: go/en/stack.go
package main

import "fmt"

/**
push() 向栈中压入一个元素
pop() 从栈顶取出一个元素
isEmpty() 判断栈是否为空
length() 获取栈中元素的数目
peer() 查询栈顶元素
*/

type Stack struct {
	// 一个切片数组
	cache []int
}

func (self Stack) length() int {
	return len(self.cache)
}

func (sk *Stack) push(n int) int {
	sk.cache = append(sk.cache, n)
	return n
}

func (sk *Stack) pop() int {
	if sk.length() == 0 {
		return 0
	}

	elem := sk.cache[sk.length()-1]
	sk.cache = sk.cache[:sk.length()-1]
	return elem
}

func (sk *Stack) isEmpty() bool {
	return sk.length() == 0
}

func (sk *Stack) peer() int {
	return sk.cache[sk.length()-1]
}

func main() {
	var sk Stack

	sk.cache = []int{1, 2, 3, 4}
	sk.push(5)

	fmt.Println(sk.cache)

	fmt.Println(sk.isEmpty())

	fmt.Println(sk.peer())
	sk.pop()
	fmt.Println(sk.length())
}
